zerver/tests/test_events.py

```diff
@@ -405,6 +405,14 @@ def realm_bot_schema(self, field_name, check):
     def do_test(self, action, event_types=None, include_subscribers=True, state_change_expected=True,
                 num_events=1):
         # type: (Callable[[], Any], Optional[List[str]], bool, bool, int) -> List[Dict[str, Any]]
+
+        '''
+        Make sure we have a clean slate of client descriptors for these tests.
+        If we don't do this, then certain failures will only manifest when you
+        run multiple tests.
+        '''
+        user_clients.clear()
+
```

Comment:
Why not just call `client.cleanup()` at the end of this function?  I think that'd be more robust, since just clearing `user_clients` will still leave other data structures in the wrong state.

Reply:
I didn't know that existed. :)

It would be nice to have a bigger-hammer reset() at the beginning of the test, though, so you don't have to rely on prior tests properly cleaning up here.  (It would be annoying to chase down a flake on a properly written test that only fails when another test doesn't clean up.)

Reply:
If you want to iterate through every `client_descriptor` in `user_clients` and call the cleanup method in this class' `setUp` method, that'd be a fine idea.  

Speaking of which, the tutorial part of this test class' existing `setUp` feels like it's actually code that should be put in exactly 1 test somewhere.  Specifically, I think `test_do_mark_hotspot_as_read` (I tested what fails if you remove it; it's just that one).